pkgs/build-support/docker/stream_layered_image.py

```diff
@@ -0,0 +1,250 @@
+function does all this.
+
+[1]: https://github.com/moby/moby/blob/master/image/spec/v1.2.md
+[2]: https://github.com/moby/moby/blob/4fb59c20a4fb54f944fe170d0ff1d00eb4a24d6f/image/spec/v1.2.md#image-json-field-descriptions
+"""  # noqa: E501
+
+
+import io
+import os
+import re
+import sys
+import json
+import hashlib
+import tarfile
+import itertools
+import threading
+from datetime import datetime
+from collections import namedtuple
+
+
+# Adds the given store paths to as a tar to the given writable stream.
+def archive_paths_to(obj, paths, mtime, add_nix, filter=None):
+    filter = filter if filter else lambda i: i
+
+    # gettarinfo makes the paths relative, this makes them
+    # absolute again
+    def append_root(ti):
+        ti.name = "/" + ti.name
+        return ti
+
+    def apply_filters(ti):
+        ti.mtime = mtime
+        return filter(ti)
+
+    def dir(path):
+        ti = tarfile.TarInfo(path)
+        ti.type = tarfile.DIRTYPE
+        return ti
+
+    with tarfile.open(fileobj=obj, mode="w|") as tar:
+        # To be consistent with the docker utilities, we need to have
+        # these directories first when building layer tarballs. But
+        # we don't need them on the customisation layer.
+        if add_nix:
+            tar.addfile(apply_filters(dir("/nix")))
+            tar.addfile(apply_filters(dir("/nix/store")))
+
+        for path in paths:
+            ti = tar.gettarinfo(os.path.join("/", path))
+            tar.addfile(apply_filters(append_root(ti)))
+
+            for root, dirs, files in os.walk(path, topdown=True):
+                for name in itertools.chain(dirs, files):
+                    name = os.path.join(root, name)
+                    ti = append_root(tar.gettarinfo(name))
+
+                    # copy hardlinks as regular files
+                    if ti.islnk():
+                        ti.type = tarfile.REGTYPE
+
+                    ti = apply_filters(ti)
+                    if ti.isfile():
+                        with open(name, "rb") as f:
+                            tar.addfile(ti, f)
+                    else:
+                        tar.addfile(ti)
+
+
+# A writable stream which only calculates the final file size and
+# sha256sum, while discarding the actual contents.
+class ExtractChecksum:
+    def __init__(self):
+        self._digest = hashlib.sha256()
+        self._size = 0
+
+    def write(self, data):
+        self._digest.update(data)
+        self._size += len(data)
+
+    def extract(self):
+        return (self._digest.hexdigest(), self._size)
+
+
+# Some metadata for a layer
+LayerInfo = namedtuple("LayerInfo", ["size", "checksum", "path", "paths"])
+
+
+# Given a list of store paths 'paths', creates a layer add append it
+# to tarfile 'tar'. Returns some a 'LayerInfo' for the layer.
+def add_layer_dir(tar, paths, mtime, add_nix=True, filter=None):
+    assert all(i.startswith("/nix/store/") for i in paths)
```

Comment:
Could you maybe add some additional info for the case where this assert might actually fail? I guess this is one of those "this should actually never happen" cases. But it would be nice to provide some useful message for when this does fail after all.

```python3
 assert all(i.startswith("/nix/store/") for i in paths), "useful message with violating path(s)"
```